vmm_core/vmm_core_defs/src/debug_rpc.rs

```diff
@@ -28,15 +29,59 @@ pub enum DebugRequest {
     /// Sets the hardware debugger state for a VP.
     SetDebugState { vp: u32, state: DebugState },
     /// Fetch the specified vp's register state.
-    GetVpState(FailableRpc<u32, Box<VpState>>),
+    GetVpState(FailableRpc<u32, Box<DebuggerVpState>>),
     /// Set the specified vp's register state.
-    SetVpState(FailableRpc<(u32, Box<VpState>), ()>),
+    SetVpState(FailableRpc<(u32, Box<DebuggerVpState>), ()>),
     /// Read from the specified GPA from the guest.
     ReadMemory(FailableRpc<(GuestAddress, usize), Vec<u8>>),
     /// Write to the specified GPA from the guest.
     WriteMemory(FailableRpc<(GuestAddress, Vec<u8>), ()>),
 }
 
+/// Register state for a VP.
+///
+/// This has all the supported architectures embedded in it to avoid having
+/// arch-specific compilation at this layer.
+#[derive(Debug, Protobuf)]
+pub enum DebuggerVpState {
+    X86_64(X86VpState),
+    Aarch64(Aarch64VpState),
```

Comment:
unfortunate that x86 has to "pay" due to ARM support being added here...

are you completely opposed to having this arch gated?

Reply:
I am not completely opposed. But aren't we trying to minimize cfg(guest_arch = ...)?

Reply:
I think my worry is that we're making different design choices based on how "impactful" the changes are.

Correct me if I'm wrong, but we use compile-time gates in things like, say, core `virt_` infrastructure, precisely because we don't want to carry x86 specific stuff on ARM (and vice-versa). We could've done something similar to this approach, where things like CPU topology would be arch-based enums, but we didn't do that.

I think whichever way we go in this code is gonna be reasonable, but nailing down some specific guidelines / philosophy around multi-arch code would be good moving forwards.

Reply:
Fair. Collecting some thoughts here... I think general there are three different approaches we can take:

1. Define types/traits/functions that are the union of the arch-specific details.
2. Define types/traits/functions that use generics to describe arch-specific details.
3. Use `cfg(guest_arch = ...)` to just gate stuff at compile time.

I think each of these have their uses, and I think we can probably be a bit more prescriptive about when you should reach for which tool.

For example, I think we generally should avoid the use of `cfg` in cross-worker communication types (like the ones defined here), because that makes it hard for us to ever support a cross-arch VMM via emulation--even if the VM worker binary needs to be built multiple times to support multiple arches, we should not require multiple builds of client binaries (such as petri).

It's also just a good principle that the build configuration should not affect the cross-process protocol--we ran into problems with petri being built with different features from openvmm and having protocol incompatibilities. It's easier to have a rule forbidding _all_ `cfg` in protocol types than to explain that it's just `feature` that's prohibited.

I also think we should generally avoid arch generics in protocol types, since generics are viral. In this case, we'd have to propagate the arch up through... `DebugRequest`, `VmProxy`, `State`, `DebuggerWorker`...

So that leaves us with unions, which I think make sense here.

But for traits and functions, different constraints are at play. There, we have experimented with `guest_arch` and generics to a greater degree. I think we'll want to revisit some of our decisions there over time, and maybe use the union technique in a few more places.

Reply:
That's a very insightful comment John, thank you. Thinking about it in terms of worker code vs. "client" code is not something I had considered, but makes a lot of sense.

Could you send a PR up which adds these thoughts to https://openvmm.dev/guide/dev_guide/contrib/code.html, so we can codify some of this thinking? Or, if you'd like to drive-by include it as part of this PR (which looks like it'll need at least one more revision anyways), that'd be great.

Reply:
I'll do this separately. Tracking via #320 so I don't forget.